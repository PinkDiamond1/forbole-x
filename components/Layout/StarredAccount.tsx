import React from 'react'
import Link from 'next/link'
import useTranslation from 'next-translate/useTranslation'
import { gql, useSubscription } from '@apollo/client'
import {
  Avatar,
  Box,
  ListItem,
  ListItemIcon,
  ListItemText,
  useTheme,
  CircularProgress,
  Typography,
} from '@material-ui/core'
import { useGeneralContext } from '../../contexts/GeneralContext'
import {
  formatCurrency,
  formatTokenAmount,
  getTotalBalance,
  getTotalTokenAmount,
  transformGqlAcountBalance,
} from '../../misc/utils'
import { CustomTheme } from '../../misc/theme'
import cryptocurrencies from '../../misc/cryptocurrencies'
import useStyles from './styles'
import { getLatestAccountBalance } from '../../graphql/queries/accountBalances'

interface StarredAccountProps {
  account: Account
}

const StarredAccount: React.FC<StarredAccountProps> = ({ account }) => {
  const themeStyle: CustomTheme = useTheme()
  const { lang } = useTranslation('common')
  const { currency } = useGeneralContext()
  const classes = useStyles()
  const crypto = cryptocurrencies[account.crypto]
  // Latest data
  const { data: balanceData, loading } = useSubscription(
    gql`
      ${getLatestAccountBalance(account.crypto)}
    `,
    { variables: { address: account.address } }
  )
  const { tokenAmounts, usdBalance } = React.useMemo(() => {
    const accountBalance = transformGqlAcountBalance(balanceData, Date.now())
    return {
      tokenAmounts: getTotalTokenAmount(accountBalance).amount,
      usdBalance: getTotalBalance(accountBalance).balance,
    }
  }, [balanceData])
  return (
    <Link key={account.address} href="/account/[address]" as={`/account/${account.address}`}>
      <ListItem className={classes.favMenuItem} button component="a">
        <ListItemIcon>
          <Avatar
            alt={crypto.name}
            src={crypto.image}
            style={{ height: themeStyle.spacing(4), width: themeStyle.spacing(4) }}
          />
        </ListItemIcon>
        <Box display="flex" flexDirection="column">
          <ListItemText
            primary={account.name}
            primaryTypographyProps={{
              variant: 'h6',
            }}
          />
          {loading ? (
            <Box>
              <CircularProgress size={20} />
            </Box>
          ) : (
            <Box>
              <Typography>{formatTokenAmount(tokenAmounts, account.crypto, lang)}</Typography>
              {/* after USD balance is added */}
              {/* <Typography>{formatCurrency(usdBalance, currency, lang)}</Typography> */}
            </Box>
          )}
        </Box>
      </ListItem>
    </Link>
  )
}

export default StarredAccount
