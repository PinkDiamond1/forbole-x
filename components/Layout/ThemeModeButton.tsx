import { IconButton } from '@material-ui/core'
import React from 'react'
import useStyles from './styles'
import DarkModeIcon from '../../assets/images/icons/icon_dark_mode.svg'
import LightModeIcon from '../../assets/images/icons/icon_light_mode.svg'
import { useSettingsContext } from '../../contexts/SettingsContext'
import useIconProps from '../../misc/useIconProps'

interface ThemeModeButtonProps {}

const ThemeModeButton: React.FC<ThemeModeButtonProps> = () => {
  const { theme, setTheme } = useSettingsContext()
  const iconProps = useIconProps(3)
  const classes = useStyles()
  const isDarkMode = theme === 'dark'
  return (
    <IconButton
      onClick={() => setTheme(isDarkMode ? 'light' : 'dark')}
      className={classes.navBarButton}
    >
      {isDarkMode ? <LightModeIcon {...iconProps} /> : <DarkModeIcon {...iconProps} />}
    </IconButton>
  )
}

export default React.memo(ThemeModeButton)
