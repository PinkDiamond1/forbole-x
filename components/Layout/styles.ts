import { makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles(
  (theme) => ({
    leftMenuContainer: {
      width: theme.spacing(32),
      position: 'fixed',
      top: 0,
      bottom: 0,
      left: 0,
      overflow: 'auto',
    },
    menu: {
      margin: theme.spacing(2, 2.5),
    },
    menuItem: {
      height: theme.spacing(6),
      marginBottom: theme.spacing(3),
      borderRadius: theme.spacing(1),
    },
    navBar: {
      position: 'fixed',
      top: 0,
      right: 0,
      left: 0,
      height: theme.spacing(11),
      display: 'flex',
      justifyContent: 'flex-end',
      alignItems: 'center',
      padding: theme.spacing(5),
    },
    navBarButton: {
      marginLeft: theme.spacing(2),
    },
    main: {
      padding: theme.spacing(3),
      paddingTop: 0,
      marginLeft: theme.spacing(32),
      marginTop: theme.spacing(11),
      display: 'flex',
      flexDirection: 'column',
    },
  }),
  { name: 'HookGlobalStyles', index: 2 }
)

export default useStyles
