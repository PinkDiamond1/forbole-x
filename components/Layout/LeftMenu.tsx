import React from 'react'
import useTranslation from 'next-translate/useTranslation'
import OverviewIcon from '../../assets/images/icons/icon_overview.svg'
import WalletManageIcon from '../../assets/images/icons/icon_wallet_manage.svg'
import DelegateIcon from '../../assets/images/icons/icon_delegate_08.svg'
import MarketIcon from '../../assets/images/icons/icon_market.svg'
import ProposalIcon from '../../assets/images/icons/icon_proposal.svg'
import AddressBookIcon from '../../assets/images/icons/icon_address_book.svg'
import ExplorerIcon from '../../assets/images/icons/icon_explorer.svg'
import Logo from '../../assets/images/logo.svg'
import { useTheme } from '@material-ui/core/styles'
import { List, ListItem, ListItemIcon, ListItemText, Paper } from '@material-ui/core'
import useStyles from './styles'
import Link from 'next/link'

interface LeftMenuProps {
  activeItem: string
}

const LeftMenu: React.FC<LeftMenuProps> = ({ activeItem }) => {
  const theme = useTheme()
  const { t } = useTranslation('common')
  const iconProps = {
    width: theme.spacing(3),
    height: theme.spacing(3),
  }
  const classes = useStyles()
  const items = React.useMemo(
    () => [
      {
        title: t('overview'),
        icon: <OverviewIcon {...iconProps} />,
        href: '/',
      },
      {
        title: t('walletManage'),
        icon: <WalletManageIcon {...iconProps} />,
        href: '/wallets',
      },
      {
        title: t('delegate'),
        icon: <DelegateIcon {...iconProps} />,
        href: '/delegate',
      },
      {
        title: t('market'),
        icon: <MarketIcon {...iconProps} />,
        href: '/market',
      },
      {
        title: t('proposals'),
        icon: <ProposalIcon {...iconProps} />,
        href: '/proposals',
      },
      {
        title: t('addressBook'),
        icon: <AddressBookIcon {...iconProps} />,
        href: '/address-book',
      },
      {
        title: t('explorer'),
        icon: <ExplorerIcon {...iconProps} />,
        href: '/explorer',
      },
    ],
    [iconProps, t]
  )

  return (
    <Paper elevation={0} className={classes.leftMenuContainer}>
      <List className={classes.menu}>
        <ListItem className={classes.menuItem} button>
          <ListItemIcon>
            <Logo
              width={theme.spacing(5)}
              height={theme.spacing(5)}
              fill={theme.palette.primary.main}
            />
          </ListItemIcon>
          <ListItemText
            primary="PORTAL"
            primaryTypographyProps={{
              variant: 'h1',
              color: 'primary',
            }}
          />
        </ListItem>
        {items.map((item) => {
          const selected = item.href === activeItem
          return (
            <Link key={item.title} href={item.href} passHref>
              <ListItem selected={selected} className={classes.menuItem} button component="a">
                <ListItemIcon>
                  {React.cloneElement(item.icon, {
                    fill: selected ? theme.palette.primary.main : theme.palette.grey[300],
                  })}
                </ListItemIcon>
                <ListItemText
                  primary={item.title}
                  primaryTypographyProps={{
                    variant: 'h6',
                    color: selected ? 'primary' : 'textSecondary',
                  }}
                />
              </ListItem>
            </Link>
          )
        })}
      </List>
    </Paper>
  )
}

export default React.memo(LeftMenu)
