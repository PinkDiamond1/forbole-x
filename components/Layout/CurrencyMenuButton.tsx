import { Button, Menu, MenuItem } from '@material-ui/core'
import React from 'react'
import currencies from '../../misc/currencies'
import useStyles from './styles'
import CurrencyIcon from '../../assets/images/icons/icon_delegate_08.svg'
import { useSettingsContext } from '../../contexts/SettingsContext'
import { IconProps } from './NavBar'

interface CurrencyMenuButtonProps extends IconProps {}

const CurrencyMenuButton: React.FC<CurrencyMenuButtonProps> = (iconProps) => {
  const { currency, setCurrency } = useSettingsContext()
  const classes = useStyles()

  const [currencyAnchor, setCurrencyAnchor] = React.useState<Element>()

  const onClose = React.useCallback(() => setCurrencyAnchor(undefined), [setCurrencyAnchor])
  const onItemClick = React.useCallback(
    (c) => {
      setCurrency(c)
      onClose()
    },
    [setCurrency, onClose]
  )

  return (
    <>
      <Button
        onClick={(e) => setCurrencyAnchor(e.currentTarget)}
        className={classes.navBarButton}
        startIcon={<CurrencyIcon {...iconProps} />}
      >
        {currency}
      </Button>
      <Menu
        anchorEl={currencyAnchor}
        getContentAnchorEl={null}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        keepMounted
        open={!!currencyAnchor}
        onClose={onClose}
      >
        {currencies
          .filter((c) => c !== currency)
          .map((c) => (
            <MenuItem key={c} onClick={() => onItemClick(c)}>
              {c}
            </MenuItem>
          ))}
      </Menu>
    </>
  )
}

export default React.memo(CurrencyMenuButton)
