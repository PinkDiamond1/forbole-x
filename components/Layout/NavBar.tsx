import { Box, IconButton } from '@material-ui/core'
import React from 'react'
import useStyles from './styles'
import NotiIcon from '../../assets/images/icons/icon_notification_32.svg'
import SettingsIcon from '../../assets/images/icons/icon_settings.svg'
import CurrencyMenuButton from './CurrencyMenuButton'
import LangMenuButton from './LangMenuButton'
import ThemeModeButton from './ThemeModeButton'
import useIconProps from '../../misc/useIconProps'

interface NavBarProps {}

const NavBar: React.FC<NavBarProps> = () => {
  const iconProps = useIconProps(3)
  const classes = useStyles()

  return (
    <Box className={classes.navBar}>
      <CurrencyMenuButton />
      <LangMenuButton />
      <IconButton className={classes.navBarButton}>
        <NotiIcon {...iconProps} />
      </IconButton>
      <ThemeModeButton />
      <IconButton className={classes.navBarButton}>
        <SettingsIcon {...iconProps} />
      </IconButton>
    </Box>
  )
}

export default React.memo(NavBar)
