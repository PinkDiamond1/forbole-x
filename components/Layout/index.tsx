import React from 'react'
import { Box } from '@material-ui/core'
import useStyles from './styles'
import LeftMenu from './LeftMenu'
import NavBar from './NavBar'

interface LayoutProps {
  activeItem: string
  children: React.ReactNode
}

const Layout: React.FC<LayoutProps> = ({ activeItem, children }) => {
  const classes = useStyles()
  return (
    <>
      <NavBar />
      <LeftMenu activeItem={activeItem} />
      <Box className={classes.main}>{children}</Box>
    </>
  )
}

export default React.memo(Layout)
