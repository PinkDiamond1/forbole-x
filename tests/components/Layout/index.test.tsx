import React from 'react'
import renderer from 'react-test-renderer'
import Layout from '../../../components/Layout'

jest.mock('../../../components/Layout/NavBar', () => 'NavBar')
jest.mock('../../../components/Layout/LeftMenu', () => 'LeftMenu')

describe('component: Layout', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <Layout activeItem="">
        <div />
      </Layout>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
