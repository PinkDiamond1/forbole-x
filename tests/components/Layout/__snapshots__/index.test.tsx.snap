// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Layout renders default state for first time user correctly 1`] = `
Array [
  <NavBar
    isChromeExt={false}
    menuWidth={32}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={true}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-19 HookGlobalStyles-main-13"
    style={
      Object {
        "marginLeft": 256,
      }
    }
  >
    <div />
  </div>,
]
`;

exports[`component: Layout renders default state for non first time user correctly 1`] = `
Array [
  <NavBar
    isChromeExt={false}
    menuWidth={32}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={true}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-25 HookGlobalStyles-main-13"
    style={
      Object {
        "marginLeft": 256,
      }
    }
  >
    <div />
  </div>,
]
`;

exports[`component: Layout renders menu collapsed state 1`] = `
Array [
  <NavBar
    isChromeExt={false}
    menuWidth={10}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={false}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-31 HookGlobalStyles-main-13"
    style={
      Object {
        "marginLeft": 80,
      }
    }
  />,
  <UnlockPasswordDialog />,
]
`;

exports[`component: Layout renders passwordRequired state for first time user correctly 1`] = `
Array [
  <NavBar
    isChromeExt={false}
    menuWidth={32}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={true}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-22 HookGlobalStyles-main-13"
    style={
      Object {
        "marginLeft": 256,
      }
    }
  >
    <GetStarted />
  </div>,
]
`;

exports[`component: Layout renders passwordRequired state for non first time user correctly 1`] = `
Array [
  <NavBar
    isChromeExt={false}
    menuWidth={32}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={true}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-28 HookGlobalStyles-main-13"
    style={
      Object {
        "marginLeft": 256,
      }
    }
  />,
  <UnlockPasswordDialog />,
]
`;
